datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -476,6 +478,26 @@ fn type_union_resolution_coercion(
                 type_union_resolution_coercion(lhs.data_type(), rhs.data_type());
             new_item_type.map(|t| DataType::List(Arc::new(Field::new("item", t, true))))
         }
+        (DataType::Struct(lhs), DataType::Struct(rhs)) => {
+            if lhs.len() != rhs.len() {
+                return None;
+            }
+
+            let types = std::iter::zip(lhs.iter(), rhs.iter())
```

Comment:
I am not familiar with the semantics, but as I read this code it would *not* coerce structs with fields named the same but in a different order. 

For example
```
{ 
  a: 1,
  b: 2
}
```
Would not be coerceable / comparable to 

```
{ 
  b: 20 // note the fields are in different order
  a: 10,
}
```

Is that intended? 🤔

Reply:
It seems like a bug 👍